### PR TITLE
[ENG-1971] Fixes spam fields

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

So the problem appears to be that the include Guid mixin is being used and that's restricting the query's to `GuidMixinQuerySet` queries, (not Node queries alone, which have more addons). So grabbing 'addons_forward` cause it to fail, because Registrations don't have forward, so that breaks the very abstracted filter query. 

## Changes

- SPAM_CHECK_FIELDS now doesn't include `addons_forward` so it doesn't check that addon every save.
- `get_spam_fields` has now been deleted, apparently not necessary and was preventing us from checking a desired field.
- Now that we can check arbitrary field, we can just limit the check for when the NodeSettings model for `forward` is saved.

## QA Notes

🐞 ! Now anything should be markable as ham without problem, reverting to the expected behavior.

⚠️ Unit tests could not replicate! ⚠️

The test app won't replicate the bug, because [this line](https://github.com/CenterForOpenScience/osf.io/blob/develop/osf/utils/requests.py#L25-L39) always "drops" the requests, leaving only a dummy_request object. Could be related to https://openscience.atlassian.net/projects/ENG/issues/ENG-1881 where the request is messed up in the test context.

## Documentation

🐞  fix, no docs.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-1971

### CODE REVIEW

File: addons/forward/models.py [hunk 1]
```diff
@@ -44,7 +44,8 @@ def save(self, request=None, *args, **kwargs):
             else:
                 user = request.user
 
-            self.owner.check_spam(user, {'addons_forward_node_settings__url'}, get_headers_from_request(request))
+            if self.owner.get_addon('forward'):
```
Comment (mfraezz) [inline addons/forward/models.py:47]: This conditional seems superfluous -- `self.owner.get_addon('forward') == self`, no?

File: osf/models/mixins.py [hunk 2]
```diff
@@ -2021,9 +2021,8 @@ def _get_spam_content(self, saved_fields):
         :param saved_fields: set
         :return: str
         """
-        spam_fields = self.get_spam_fields(saved_fields)
```
Comment (mfraezz) [inline osf/models/mixins.py:2024]: There are some fields that shouldn't be spam-checked, and it makes more sense to get only spam fields here than, say, at every call to `check_spam`.

File: addons/forward/models.py [hunk 3]
```diff
@@ -44,7 +44,7 @@ def save(self, request=None, *args, **kwargs):
             else:
                 user = request.user
 
-            self.owner.check_spam(user, {'addons_forward_node_settings__url'}, get_headers_from_request(request))
+            self.owner.check_spam(user, {'addons_forward_node_settings__url'}, get_headers_from_request(request), force=True)
```
Comment (mfraezz) [inline addons/forward/models.py:47]: As per the last review, how does `force` address the following?

> cases where the `url` is added when the node is private, but isn't checked when it becomes public.

Comment (Johnetordoff) [reply addons/forward/models.py:47]: Presumably that situation never produces spam because this happen on every model save regardless of whether it's public or private, but I see now this is broken by `if settings.SPAM_CHECK_PUBLIC_ONLY and not self.is_public:` but I should be able to override that one line and be good. I can also add something on ln 46 that checks if `owner.is_public`, hopefully that's on the right track.

File: osf/models/mixins.py (hunk 2 above)
Comment (Johnetordoff) [reply osf/models/mixins.py:2024]: yeah I meant to revert this, sorry!

File: addons/forward/models.py [hunk 4]
```diff
@@ -44,7 +44,7 @@ def save(self, request=None, *args, **kwargs):
             else:
                 user = request.user
 
-            self.owner.check_spam(user, {'addons_forward_node_settings__url'}, get_headers_from_request(request))
+            self.owner.do_check_spam(user, {'addons_forward_node_settings__url'}, get_headers_from_request(request), force=True)
```
Comment (mfraezz) [inline addons/forward/models.py:47]: This should not be `do_check_spam`; the signature is `(self, author, author_email, content, request_headers, update=True)`.

Also, `check_spam` handles `SPAM_CHECK_PUBLIC_ONLY`, `do_` does not.

File: addons/forward/models.py (hunk 3 above)
Comment (mfraezz) [reply addons/forward/models.py]: Bypassing the `SPAM_CHECK_PUBLIC_ONLY` check seems like the wrong path.

Why not just conditionally include `'addons_forward_node_settings__url'` in `get_spam_fields` `if node.has_addon('forward')`?

Comment (Johnetordoff) [reply addons/forward/models.py]: > Why not just conditionally include 'addons_forward_node_settings__url' in get_spam_fields if node.has_addon('forward') ?

Yeah this makes sense, probably better. 

Just side comment though, as for `SPAM_CHECK_PUBLIC_ONLY` I don't see why this behavior is configurable here at all, our code seems to assume this is True implictly, like in `get_spam_fields`,

**Comment by brianjgeiger** [comment]
So it sounds like we either need to do what the ticket originally suggested, or ensure that on all saves, we check if there's a forward addon and, if so, add its url to the content that we're checking. Does that sound about right?

File: osf/models/node.py [hunk 5]
```diff
@@ -1140,8 +1140,14 @@ def visible_contributors(self):
...
-        return self.SPAM_CHECK_FIELDS if self.is_public and 'is_public' in saved_fields else self.SPAM_CHECK_FIELDS.intersection(
-            saved_fields)
+        if self.is_public and 'is_public' in saved_fields:
+            if self.get_addon('forward'):
+                return self.SPAM_CHECK_FIELDS
+            else:
+                return self.SPAM_CHECK_FIELDS.difference('addons_forward_node_settings__url')
+
+        else:
+            return self.SPAM_CHECK_FIELDS.intersection(saved_fields)
```
Comment (mfraezz) [inline osf/models/node.py:1150]: Difference accepts a `set`.
```
check_fields = self.SPAM_CHECK_FIELDS.copy()
if not self.has_addon('forward'):
    check_fields.remove('addons_forward_node_settings__url')
return fields if self.is_public and 'is_public' in saved_fields else check_fields.intersection(saved_fields)
```

File: addons/forward/models.py [hunk 6]
```diff
@@ -35,17 +32,6 @@ def after_register(self, node, registration, user, save=True):
 
         return clone, None
 
-    def save(self, request=None, *args, **kwargs):
```
Comment (mfraezz) [inline addons/forward/models.py:38]: Why was this removed? If it goes from `not spammy url` to `spammy url`, we probably still want to be sniffing it.

File: addons/forward/models.py (hunk 3 above)
Comment (mfraezz) [reply addons/forward/models.py]: >`if self.is_public and 'is_public' in saved_fields`

This causes it to check all potentially-spammy fields when something goes `private->public`, thereafter checking only changed fields.

File: osf/models/mixins.py [hunk 7]
```diff
@@ -2013,14 +2013,6 @@ def confirm_ham(self, save=False):
             self.save()
 
     def _get_spam_content(self, saved_fields):
-        """
-        This function retrieves retrieves strings of potential spam from various DB fields. Also here we can follow
```
Comment (mfraezz) [inline osf/models/mixins.py:2017]: Why was this docstring removed?

File: osf/models/node.py (hunk 5 above)
Comment (Johnetordoff) [reply osf/models/node.py:1150]: > Difference accepts a set.

It also accepts a string!  
```
In [215]: {'a', 'b', 'c'}.difference('a').copy()
Out[215]: {'b', 'c'}
In [215]: {'a', 'b', 'c'}.difference('a').copy()
Out[215]: {'b', 'c'}

It's the copy method I don't understand here, but I put in the new commit.

Comment (mfraezz) [reply osf/models/node.py]: It does accept a string, because strings can be reduced to a `set` of characters.
```
In [8]: Node.SPAM_CHECK_FIELDS.difference('addons_forward_node_settings__url')
Out[8]: {'addons_forward_node_settings__url', 'description', 'title'}

In [9]: set('addons_forward_node_settings__url')
Out[9]: {'_', 'a', 'd', 'e', 'f', 'g', 'i', 'l', 'n', 'o', 'r', 's', 't', 'u', 'w'}

In [10]: set({'addons_forward_node_settings__url'})
Out[10]: {'addons_forward_node_settings__url'}

In [11]: Node.SPAM_CHECK_FIELDS.difference({'addons_forward_node_settings__url'})
Out[11]: {'description', 'title'}
```

Comment (mfraezz) [reply osf/models/node.py]: >It's the copy method I don't understand here, but I put in the new commit.

Shallow `copy` of the `SPAM_CHECK_FIELDS` set, otherwise this happens:
```
In [12]: fields = Node.SPAM_CHECK_FIELDS

In [13]: Node.SPAM_CHECK_FIELDS
Out[13]: {'addons_forward_node_settings__url', 'description', 'title'}

In [14]: fields
Out[14]: {'addons_forward_node_settings__url', 'description', 'title'}

In [15]: fields.remove('addons_forward_node_settings__url')

In [16]: fields
Out[16]: {'description', 'title'}

In [17]: Node.SPAM_CHECK_FIELDS
Out[17]: {'description', 'title'}
```

File: osf/models/mixins.py [hunk 8]
```diff
@@ -2013,18 +2013,13 @@ def confirm_ham(self, save=False):
...
-        :param saved_fields: set
-        :return: str
-        """
         spam_fields = self.get_spam_fields(saved_fields)
         content = []
+        print(self)
+
         for field in spam_fields:
             exclude_null = {field + '__isnull': False}
+            print(field)
```
Comment (mfraezz) [inline osf/models/mixins.py:2022]: Please remove `print` statements.

File: osf/models/mixins.py (hunk 7 above)
Comment (Johnetordoff) [reply osf/models/mixins.py:2017]: I meant to remove the docstring on check_spam

File: osf/models/node.py (hunk 5 above)
Comment (Johnetordoff) [reply osf/models/node.py]: That's what I get for testing with a single character.
